### PR TITLE
20087 - Statement locking and notification

### PR DESCRIPTION
**Pull request by seeker25**
*Issue #:*
https://github.com/bcgov/entity/issues/20087

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

File: jobs/payment-jobs/tasks/statement_due_task.py [hunk 1]
```diff
@@ -79,27 +79,34 @@ def _notify_for_monthly(cls):
...
         for payment_account in eft_payment_accounts:
             try:
-                statement = cls._find_most_recent_statement(
-                    payment_account.auth_account_id, StatementFrequency.MONTHLY.value)
+                if not (statement := cls._find_most_recent_statement(
+                        payment_account.auth_account_id, StatementFrequency.MONTHLY.value)):
+                    continue
                 action, due_date = cls._determine_action_and_due_date_by_invoice(statement.id)
                 total_due = Statement.get_summary(payment_account.auth_account_id, statement.id)['total_due']
-                if action and total_due > 0 and (emails := cls._determine_recipient_emails(statement, action)):
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/statement_due_task.py:86]: Previously if there was no email, it wouldn't lock

File: jobs/payment-jobs/tasks/statement_due_task.py [hunk 2]
```diff
@@ -79,27 +79,34 @@ def _notify_for_monthly(cls):
...
+                if action and total_due > 0:
+                    emails = cls._determine_recipient_emails(statement, action)
                     if action == StatementNotificationAction.OVERDUE:
                         current_app.logger.info('Freezing payment account id: %s and locking auth account id: %s',
                                                 payment_account.id, payment_account.auth_account_id)
                         AuthEvent.publish_lock_account_event(payment_account)
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/statement_due_task.py:92]: Feature flag the lock

File: jobs/payment-jobs/tasks/statement_due_task.py [hunk 3]
```diff
@@ -79,27 +79,34 @@ def _notify_for_monthly(cls):
         current_app.logger.info(f'Processing {len(eft_payment_accounts)} EFT accounts for monthly reminders.')
```
Comment (seeker25) [inline jobs/payment-jobs/tasks/statement_due_task.py:79]: Change update_invoice_overdue_status so we get a list of invoice ids and add them to the nsf invoices table, so they show up in the NSF unlock screen for EFT

**Comment by seeker25** [comment]
Just taking this for a test drive now.

File: pay-api/src/pay_api/models/invoice.py [hunk 4]
```diff
@@ -37,6 +38,18 @@
...
+def determine_overdue_date(context):
+    """Determine the overdue date with the correct time offset."""
+    created_on = context.get_current_parameters()['created_on']
+    target_date = created_on.date() + relativedelta(months=2,
+                                                    day=1)
+    target_datetime = datetime.combine(target_date, datetime.min.time())
+    # Correct for daylight savings.
+    hours = target_datetime.astimezone(pytz.timezone('America/Vancouver')).utcoffset().total_seconds() / 60 / 60
+    target_date = target_datetime.replace(tzinfo=timezone.utc) + relativedelta(hours=-hours)
+    return target_date.replace(tzinfo=None)
```
Comment (seeker25) [inline pay-api/src/pay_api/models/invoice.py:50]: If i didn't do this, I would see weird behavour, would sometimes remove the hours.. but in the model unit test the model persists. tzinfo=None seems to make it always show the hours

File: pay-api/src/pay_api/models/non_sufficient_funds.py [hunk 5]
```diff
@@ -21,7 +21,7 @@
 from .db import db
 
 
-class NonSufficientFundsModel(BaseModel):  # pylint: disable=too-many-instance-attributes
+class NonSufficientFunds(BaseModel):  # pylint: disable=too-many-instance-attributes
```
Comment (seeker25) [inline pay-api/src/pay_api/models/non_sufficient_funds.py:24]: renamed to make this consistent

File: pay-api/src/pay_api/models/non_sufficient_funds.py [hunk 6]
```diff
@@ -51,6 +51,11 @@ class NonSufficientFundsModel(BaseModel):  # pylint: disable=too-many-instance-a
     invoice_id = db.Column(db.Integer, ForeignKey('invoices.id'), nullable=False)
     invoice_number = db.Column(db.String(50), nullable=False, index=True, comment='CFS Invoice number')
 
+    @classmethod
+    def find_by_invoice_id(cls, invoice_id: int):
```
Comment (seeker25) [inline pay-api/src/pay_api/models/non_sufficient_funds.py:55]: small helper function only works on NonSufficientFunds (no complex joins to other entity types, should be fine)

**Review by ochiu — APPROVED** [review]
Looks good

**Pull request merged**